ls1intum/Artemis
# `Lectures`: Add visible date to lectures

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


Lectures are always visible to students, even if they are not relevant yet.

### Description

The lecture table now contains an additional column: visible_date
It can be set by users in the same way as the start and end dates through the client.
When students request the lectures from the server, they are filtered for visibility.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Course with messaging enabled

1. Login as an instructor
2. Create a new lecture with visible date
3. Confirm that the visible cannot be set after the start date
4. Import the newly created lecture and confirm that the visible date is the same as for the original lecture
5. Set the visible date to the future
6. Login as a student and confirm that the lecture is invisible
7. Confirm that the lecture channel is also invisible
8. Change the visible to the past and confirm that the lecture is visible to the student
9. Confirm that the lecture channel is visible to the student

### Review Progress






#### Code Review
#### Manual Tests

### Test Coverage







### Screenshots


When creating a new lecture, there is a new input field for the visible date:
<img width="1250" alt="input field" src="https://github.com/ls1intum/Artemis/assets/44754405/79f337a2-e680-415b-bb09-8cc2c3c8968b">

The visible date value is shown in the overview after import/create/edit:
<img width="306" alt="lecture overview" src="https://github.com/ls1intum/Artemis/assets/44754405/ecfa0a08-49b8-4b19-92df-570245598a4e">

And it is shown in the overview list of lectures in an additional column:
<img width="1687" alt="lectures list" src="https://github.com/ls1intum/Artemis/assets/44754405/fe001022-6b02-43b1-aa95-3c09dba56e37">

For students, lectures with a visible date in the future are hidden:
<img width="1689" alt="hidden lecture" src="https://github.com/ls1intum/Artemis/assets/44754405/4eac5813-2c73-4efa-8bcb-198c462a2146">

As soon as the visible date passes, the lecture becomes visible:
<img width="1689" alt="visible lecture" src="https://github.com/ls1intum/Artemis/assets/44754405/8a8179c8-8641-4f4f-bf70-19cb69c9f682">
